### PR TITLE
Updated `EmbeddingsWriter` to support multi-embedding file outputs

### LINKED ISSUE #376: Improve grouping feature in `EmbeddingsWriter`
For slide level tasks, we should use `group_key` to group patches from the same slide into the same .pt file(s).
The way it is implemented now still generates one file per patch. We probably should accumulate patches in memory over multiple batches before saving them to file.

### PR DESCRIPTION
**Pull request by nkaenzig**
Closes #376 

- Implemented support for adding multiple embeddings to the same `.pt` if they share the same filename (e.g. slide_id)
- Made `EmbeddingsWriter` agnostic to WSI's by replacing slide_id field extraction by generic `metadata_keys`
- Refactored & updated unit tests

This helps to prevent the explosion of the number of files for slide level tasks, where you can have thousands of patch embeddings per slide. Each slide now results in a single `.pt` file of shape `[n_patches, embedding_dim]`.

### CODE REVIEW

File: src/eva/core/callbacks/writers/embeddings.py [hunk 1]
```diff
@@ -122,38 +125,105 @@ def _get_embeddings(self, prediction: torch.Tensor) -> torch.Tensor:
...
+    write_queue: multiprocessing.Queue,
+    output_dir: str,
+    metadata_keys: List[str],
+    save_every_n: int,
+    overwrite: bool = False,
 ) -> None:
-    manifest_file, manifest_writer = _init_manifest(output_dir, overwrite)
+    manifest_file, manifest_writer = _init_manifest(output_dir, metadata_keys, overwrite)
+
+    save_name_to_items: Dict[str, ITEM_DICT_ENTRY] = {}
```
Comment (roman807) [inline src/eva/core/callbacks/writers/embeddings.py:156]: not sure if `save_name_to_items` is the best name. It suggests you "save a name to a list of items.."

Can't think of a very good name, but how about `items_dict` or `items_to_save`?

Comment (nkaenzig) [reply src/eva/core/callbacks/writers/embeddings.py:156]: I like to use variable names in dictionaries that describe the actual mapping, but indeed the `save_` prefix makes it confusing here. What about `name_to_items`?

File: src/eva/core/callbacks/writers/embeddings.py [hunk 2]
```diff
@@ -122,38 +125,105 @@ def _get_embeddings(self, prediction: torch.Tensor) -> torch.Tensor:
...
 
-def _save_prediction(prediction_buffer: io.BytesIO, save_name: str, output_dir: str) -> None:
-    save_path = os.path.join(output_dir, save_name)
-    prediction = torch.load(io.BytesIO(prediction_buffer.getbuffer()), map_location="cpu")
+def _save_items(
+    save_name_to_items: Dict[str, ITEM_DICT_ENTRY],
+    metadata_keys: List[str],
+    output_dir: str,
+    manifest_writer: Any,
+) -> Dict[str, ITEM_DICT_ENTRY]:
```
Comment (roman807) [inline src/eva/core/callbacks/writers/embeddings.py:189]: even though this is wip, i would already add a docstring with description. The logic (together with `save_predictions`) is quite tricky and a concise description might make it easier to follow.

File: src/eva/core/callbacks/writers/embeddings.py [hunk 3]
```diff
@@ -122,38 +125,105 @@ def _get_embeddings(self, prediction: torch.Tensor) -> torch.Tensor:
...
+    for save_name, entry in save_name_to_items.items():
+        if len(entry.items) > 0:
+            save_path = os.path.join(output_dir, save_name)
+            is_first_save = entry.save_count == 0
+            if is_first_save:
+                _, target, input_name, _, split, metadata = QUEUE_ITEM(*entry.items[0])
+                metadata = [metadata[key] for key in metadata_keys]  # type: ignore
+                _update_manifest(target, input_name, save_name, split, metadata, manifest_writer)
+            else:
+                pass
```
Comment (roman807) [inline src/eva/core/callbacks/writers/embeddings.py:199]: remove the 2 lines above?

File: src/eva/core/callbacks/writers/embeddings.py (hunk 1 above)
Comment (roman807) [reply src/eva/core/callbacks/writers/embeddings.py:156]: thats definitely better. Do we follow the <key>_to_<value> naming pattern elsewhere in eva?

Comment (nkaenzig) [reply src/eva/core/callbacks/writers/embeddings.py:156]: hehe good question, at least I mostly try to use the `to` pattern :D

Comment (nkaenzig) [reply src/eva/core/callbacks/writers/embeddings.py:156]: done

File: src/eva/core/callbacks/writers/embeddings.py (hunk 3 above)
Comment (nkaenzig) [reply src/eva/core/callbacks/writers/embeddings.py:199]: good catch, removed

File: src/eva/core/callbacks/writers/embeddings.py (hunk 2 above)
Comment (nkaenzig) [reply src/eva/core/callbacks/writers/embeddings.py:188]: good point, I now added a couple of docstrings to explain what happens.

**Review by roman807 — APPROVED** [review]
Great thanks, i think its good to merge

**Review by ioangatop — APPROVED** [review]
Looks great!

**Pull request merged**